packages/medusa/src/services/gift-card.ts

```diff
@@ -0,0 +1,317 @@
+          id: result.id,
+        })
+
+      return result
+    })
+  }
+
+  /**
+   * Gets a gift card by id.
+   * @param giftCardId - id of gift card to retrieve
+   * @param config - optional values to include with gift card query
+   * @return the gift card
+   */
+  async retrieve(
+    giftCardId: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const query: FindOneOptions<GiftCard> = {
+        where: { id: giftCardId },
+      }
+
+      if (config.select) {
+        query.select = config.select
+      }
+
+      if (config.relations) {
+        query.relations = config.relations
+      }
+
+      const rels = query.relations as Array<keyof GiftCard>
+      delete query.relations
+
+      const giftCard = await giftCardRepo.findOneWithRelations(rels, query)
+
+      if (!giftCard) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Gift card with ${giftCardId} was not found`
+        )
+      }
+
+      return giftCard
+    })
+  }
+
+  async retrieveByCode(
+    code: string,
+    config: FindConfig<GiftCard> = {}
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const query: FindOneOptions<GiftCard> = {
+        where: { code },
+      }
+
+      if (config.select) {
+        query.select = config.select
+      }
+
+      if (config.relations) {
+        query.relations = config.relations
+      }
+
+      const rels = query.relations as Array<keyof GiftCard>
+      delete query.relations
+
+      const giftCard = await giftCardRepo.findOneWithRelations(rels, query)
+
+      if (!giftCard) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Gift card with ${code} was not found`
+        )
+      }
+
+      return giftCard
+    })
+  }
+
+  /**
+   * Updates a giftCard.
+   * @param giftCardId - giftCard id of giftCard to update
+   * @param update - the data to update the giftCard with
+   * @return the result of the update operation
+   */
+  async update(
+    giftCardId: string,
+    update: UpdateGiftCardInput
+  ): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      const giftCard = await this.retrieve(giftCardId)
+
+      const { region_id, metadata, balance, ...rest } = update
+
+      if (region_id && region_id !== giftCard.region_id) {
+        const region = await this.regionService_.retrieve(region_id)
+        giftCard.region_id = region.id
+      }
+
+      if (metadata) {
+        giftCard.metadata = await setMetadata(giftCard, metadata)
+      }
+
+      if (typeof balance !== "undefined") {
+        if (balance < 0 || giftCard.value < balance) {
```

Comment:
Suggestion : should we combine the if conditions and move out the assignment ?

Reply:
The inner "if" only applies if the balance is not undefined which is the only case where we also want to assign the balance. If we move the assignment out we have to assign with a ternary or in another if statement.